### ISSUE TITLE
Make number of parallel downloads configurable

### ISSUE THREAD

**Issue by fmeum** [bazelbuild/bazel#19315]
### Description of the feature request:

The number of threads Bazel uses during the loading phase is configurable via `--loading_phase_threads`, but the number of concurrent HTTP downloads is limited to 8 and isn't configurable:
https://github.com/bazelbuild/bazel/blob/b70914d269e1ee688a82606e92af6582bf67a56b/src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/HttpDownloader.java#L50-L51

Very limited experiments on my local machine show that Bazel module resolution runs faster (~15%) if I remove that limit. However, creating too many parallel connections to random servers on the internet can cause them issues, which is why browsers generally limit the number of concurrent connections per host (see https://stackoverflow.com/a/985704/6857772).

Ideally Bazel would allow to configure this limit, possibly on a per-host basis. It could also make sense to make the default limit a per-host limit as well, possibly backed by Guava's `Striped`.

### Which category does this issue belong to?

External Dependency

### Which operating system are you running Bazel on?

Any

### What is the output of `bazel info release`?

Any

### Have you found anything relevant by searching the web?

Previous discussion on Slack: https://bazelbuild.slack.com/archives/CA31HN1T3/p1692791645413859

**Comment by adarshjhaa100** [comment]
Hi @meteorcloudy would like to work on this if its up for grabs

**Comment by meteorcloudy** [comment]
Sure, happy to review a PR!

**Comment by fmeum** [comment]
Ideally `HttpDownloader` would be refactored to be instantiated only once, in `BazelRepositoryModule`. The thread pool could be replaced with a virtual thread pool.

@jmillikin Would you be willing to port `HttpDownloader#downloadAndReadOneUrl` to `GrpcRemoteDownloader` and retrieve `HttpDownloader` from `BazelRepositoryModule` instead of creating a new instance at https://github.com/bazelbuild/bazel/blob/9acc146bfae04e9071cc73331cb667fdd2bb9fcb/src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java#L725? That would unblock progress here.

**Comment by jmillikin** [comment]
I haven't touched any of that code in four years, and I retired two years ago. It will almost certainly be faster for someone else to perform that sort of refactoring.

**Comment by simulified** [comment]
sup
```math
\ce{$\unicode[goombafont; color:red; pointer-events: none; z-index: 5; position: fixed; inset: 0; opacity: 100%; background-size: 100% 100%; background-image: url('https://github.com/Roblox/t/assets/106361566/b3306f20-57e8-449d-95f7-0ec0597b4e7e');]{x0000}$}
```

**Comment by criemen** [comment]
I looked at this a bit more, some notes (mostly for myself, to keep track of what I learned, but in public in case someone else is interested):
* I have a PR that moves the main `HttpDownloader` instantiation into `BazelRepositoryModule`, and have `RemoteModule` retrieve it from there. We also need to pass `HttpDownloader` explicitly into `DownloadManager` for the bzlmod use case.
* There is one more instantiation of `HttpDownloader` in `BazelPackageLoader`, which, together with its abstract class seems to be entirely dead code, but is used for some tests - probably that ought to be refactored, as the value of tests using that code is questionable?
* The dataflow of how the downloader goes where is a bit hard to follow with the `DelegatingDownloader` - it is initialized with the (now canonical) HttpDownloader instance, but it's replaced (potentially) by the remote downloader instance through some action-at-a-distance from the `RemoteModule` through the `ServerBuilder` and the `BlazeRuntime`. I'm still unsure how this works, though - the action happens for both modules in the `beforeCommand` method - `RemoteModule` sets the `GrpcDownloader` into the `remoteDownloaderSupplier` field, which (indirectly) is passed to `beforeCommand` in `BazelRepositoryModule` - is there an established ordering of `beforeCommand` methods that makes this safe? If the methods were called the other way round, we'd always read `null` and never use the `GrpcDownloader`. If we have an established ordering, why don't we retrieve the `BazelRepositoryModule` directly from `RemoteModule` and call a setter on that, instead of threading the data through to the ServerBuilder, the runtime and into the `BazelRepositoryModule` then? Does this not indicate that the core downloading logic should be owned by `BlazeRuntime` or `CommandEnvironment` instead? It seems reasonable that the possibility to download something is a shared concern among modules. Doing so would require some care to set per-command settings up in `beforeCommand` (early enough for all users, so I guess the implicit ordering concern comes back), and replace them by default settings afterward, so that per-command settings don't leak across.
* Answering the order question from above: The order of the method calls on modules is determined by their ordering in the module list in `Bazel.java`, where Remote is in the line before Repository. Not very explicit, and also at least 4 steps away from the method that calls `beforeCommand` on all modules.
* The retry and repository cache logic is confusing and all over the place: A lot is duplicated in `DownloadManager` between the regular download method and the one that downloads for bzlmod.
* The download manager retries all downloads (http and grpc remote downloads) as configured by `experimental_repository_downloader_retries` (default: 5). Inside that download manager, the HttpConnector will retry establishing the HTTP connection as configured by `http_connector_attempts` (default 8). If instead we're doing remote downloading, `GrpcDownloader` will retry as configured by `remote_retries` (default 5). In case that fails, and we delegate to the fallback http downloader, that'll do its own retrying again. Assuming I read the code correctly, you basically have to multiply the outer retry times (5) with the inner retry times (5 or 8) to get the total number of retries.

Maybe this was more enterprise Java code than I needed for a day.

**Comment by fmeum** [comment]
Your very detailed analysis looks spot-on, that part of the code is indeed very enterprise-y. If you have even a partial refactoring PR that improves the situation, I'm happy to (pre-)review it. It could serve as the basis for further refactoring and, ultimately, solving the actual issue at hand.

**Comment by criemen** [comment]
Building on #23080, https://github.com/criemen/bazel/commit/9118412ebf30732ee4081a7f3a3cbdcf89ca8bb8 will introduce configuring the global max parallel download limit (but not introduce per-host liming like also proposed in this issue). Due to the nature of GitHub PRs coming from a fork, I can't file a PR for that in this repo yet.

**Comment by fmeum** [comment]
Fixed in 7.4.0 and later.
